hazelcast/db.py

```diff
@@ -0,0 +1,509 @@
+        self._rownumber = -1
+        self._closed = False
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def __iter__(self) -> Optional[Iterator[SqlRow]]:
+        return self._iter
+
+    @property
+    def connection(self):
+        return self._conn
+
+    @property
+    def description(self) -> Union[List[DescriptionColumn], None]:
+        return self._description
+
+    @property
+    def rowcount(self) -> int:
+        return -1
+
+    @property
+    def rownumber(self) -> Optional[int]:
+        if self._rownumber < 0:
+            return None
+        return self._rownumber
+
+    def close(self):
+        if not self._closed:
+            self._closed = True
+            self._conn._close_cursor(self)
+            if self._res:
+                self._res.close()
+                self._res = None
+
+    def execute(self, operation: str, params: Optional[Tuple] = None) -> None:
+        if params is not None and not isinstance(params, tuple):
+            raise InterfaceError("params must be a tuple or None")
+        params = params or ()
+        self._ensure_open()
+        self._rownumber = -1
+        self._iter = None
+        self._res = None
+        cbs = _DEFAULT_CURSOR_BUFFER_SIZE
+        if self.arraysize > 0:
+            cbs = self.arraysize
+        self._description = None
+        res = (
+            self._conn._get_client()
+            .sql.execute(operation, *params, cursor_buffer_size=cbs)
+            .result()
+        )
+        if res.is_row_set():
+            self._rownumber = 0
+            self._res = res
+            self._description = self._make_description(res.get_row_metadata())
+            self._iter = res.__iter__()
+
+    def executemany(self, operation: str, seq_of_params: Sequence[Tuple]) -> None:
+        self._ensure_open()
+        self._rownumber = -1
+        self._iter = None
+        self._res = None
+        futures = []
+        svc = self._conn._get_client().sql
+        for params in seq_of_params:
+            futures.append(
+                svc.execute(
+                    operation, *params, expected_result_type=SqlExpectedResultType.UPDATE_COUNT
+                )
+            )
+        for fut in futures:
+            fut.result()
+
+    def fetchone(self) -> Optional[SqlRow]:
+        if self._iter is None:
+            raise InterfaceError("fetch can only be called after row returning queries")
+        if self._rownumber is None:
+            self._rownumber = 0
+        try:
+            row = next(self._iter)
+            self._rownumber += 1
+            return row
+        except StopIteration:
+            return None
+
+    def fetchmany(self, size: Optional[int] = None) -> List[SqlRow]:
+        if self._iter is None:
+            raise InterfaceError("fetchmany can only be called after row returning queries")
+        if self._rownumber is None:
```

Comment:
when can row_number be none?

Reply:
Leftover, removed at: https://github.com/hazelcast/hazelcast-python-client/pull/586/commits/6f0929e54cd6bac8cc755b1bae86fc73245df4a1